lib/fast_jsonapi/object_serializer.rb

```diff
@@ -149,11 +159,18 @@ def cache_options(cache_options)
 
       def attributes(*attributes_list, &block)
         attributes_list = attributes_list.first if attributes_list.first.class.is_a?(Array)
+        options = attributes_list.last.is_a?(Hash) ? attributes_list.pop : {}
         self.attributes_to_serialize = {} if self.attributes_to_serialize.nil?
+        self.optional_attributes_to_serialize = {} if self.optional_attributes_to_serialize.nil?
+
         attributes_list.each do |attr_name|
           method_name = attr_name
           key = run_key_transform(method_name)
-          attributes_to_serialize[key] = block || method_name
+          if options[:if].present?
+            optional_attributes_to_serialize[key] = [method_name, options[:if]]
+          else
+            attributes_to_serialize[key] = block || method_name
+          end
```

Comment:
I think it will be more maintainable if we put all attributes into a single `attributes_to_serialize` list, rather than splitting them up.  It will help to have one single code path to enumerate the attributes and to evaluate them, as we continue to evolve this and add more option parameters besides `if`.  The `if` check should be part of the evaluation process.

Reply:
Good points. I like the idea of the `AttributeDefinition` class if it's to have a defined structure. Using a hash and expecting it to be of a particular form can get fragile as multiple people continue iterating on it (in my opinion).

Reply:
@TrevorHinesley Yeah, in fact, the `AttributeDefinition` class can be the one with the serialization method:

```ruby
class AttributeDefinition
  ...
  def serialize(record, serialization_params, output_hash)
    if include_attribute?(record, serialization_params)
      output_hash[key] = ...
    end

  def include_attribute?(record, serialization_params)
    if conditional_proc.present?
      conditional_proc.call(record, serialization_params)
    else
      true
    end
  end
end
```

Reply:
@christophersansone re: EDIT -- totally understand, but since we have the ability to use blocks for custom methods as it stands, it seems reasonable that this could be merged in, then #49 could be updated to work with this new functionality as necessary (whether that means rolling this back or revising its implementation a bit). 

I don't mean to imply that we should just ignore other PRs as these kinds of abstractions are discussed, but I don't think it should be a primary concern for this PR if it makes the most sense for this feature set (considering the other PR isn't merged in, and the focus of this PR is implementing this feature into the existing codebase). Hope that makes sense.

Reply:
@TrevorHinesley 👍 Totally agree, let's focus on the right implementation here and worry about #49 later.  (It could be as simple as passing the serializer instance to `AttributeDefinition#serialize` and having it be called... lots of options here, not worth fretting about now though.)